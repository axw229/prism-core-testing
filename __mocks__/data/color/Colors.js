// @flow
import cloneDeep from 'lodash/cloneDeep'
import sample from 'lodash/sample'

import type { ColorSetPayload, Color } from '../../../src/shared/types/Colors'

export const Colors: ColorSetPayload = {
  "Red": [
    [
      {
        "colorNumber": "6561",
        "coordinatingColors": {
          "coord2ColorId": 1926,
          "coord1ColorId": 2250,
          "whiteColorId": 2681
        },
        "description": [
          "Subdued",
          "Muted",
          "Brilliant",
          "Radiant"
        ],
        "id": "2248",
        "isExterior": true,
        "isInterior": true,
        "name": "Teaberry",
        "rgb": 15454683,
        "lrv": 68.551,
        "colorFamilyNames": [
          "Red"
        ],
        "brandKey": "SW",
        "red": 235,
        "green": 209,
        "blue": 219,
        "hue": 0.935897435897436,
        "saturation": 0.39393939393939387,
        "lightness": 0.8705882352941177,
        "hex": "#ebd1db",
        "isDark": false,
        "cssrgb": "rgb(235,209,219)",
        "storeStripLocator": "102-C1",
        "similarColors": [
          "2255",
          "2262",
          "2269",
          "2662",
          "2535",
          "2276",
          "2658",
          "1974",
          "2543",
          "2531"
        ]
      },
      {
        "colorNumber": "6568",
        "coordinatingColors": {
          "coord2ColorId": 11244,
          "coord1ColorId": 2687,
          "whiteColorId": 2681
        },
        "description": [
          "Subdued",
          "Muted",
          "Brilliant",
          "Radiant"
        ],
        "id": "2255",
        "isExterior": true,
        "isInterior": true,
        "name": "Lighthearted Pink",
        "rgb": 15586781,
        "lrv": 70.758,
        "brandedCollectionIds": [
          31797,
          31791
        ],
        "brandedCollectionNames": [
          "Pottery Barn Fall/Winter 2018",
          "Pottery Barn Fall/Winter 2018"
        ],
        "colorFamilyNames": [
          "Red"
        ],
        "brandKey": "SW",
        "red": 237,
        "green": 213,
        "blue": 221,
        "hue": 0.9444444444444445,
        "saturation": 0.3999999999999999,
        "lightness": 0.8823529411764706,
        "hex": "#edd5dd",
        "isDark": false,
        "cssrgb": "rgb(237,213,221)",
        "storeStripLocator": "103-C1",
        "similarColors": [
          "2248",
          "2262",
          "2269",
          "2276",
          "1974",
          "1981",
          "2658",
          "1995",
          "2662",
          "1988"
        ]
      }
    ]
  ]
}

export function getColors (): ColorSetPayload {
  return cloneDeep(Colors)
}

export function getColor (): Color | void {
  const colors = getColors()
  const fam1 = Object.keys(colors)[0]
  if (colors[fam1]) {
    return sample(colors[fam1][0])
  }
}
