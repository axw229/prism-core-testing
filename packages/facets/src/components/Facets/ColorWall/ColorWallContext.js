// @flow
import React from 'react'
import noop from 'lodash/noop'
import type { Color } from 'src/shared/types/Colors.js.flow'
import { type GridBounds } from './ColorWall.flow'

export type ColorWallA11yContextProps = {
  a11yFocusCell: number[] | void,
  a11yFocusChunk: GridBounds,
  a11yFromKeyboard: boolean,
  a11yFocusOutline: boolean,
  a11yMaintainFocus: boolean
}

export const colorWallA11yContextDefault: ColorWallA11yContextProps = {
  a11yFocusCell: undefined,
  a11yFocusChunk: undefined,
  a11yFromKeyboard: false,
  a11yFocusOutline: false,
  a11yMaintainFocus: false
}

export type ColorWallContextProps = {
  activeColorRouteBuilderRef?: { current: (Color) => void },
  addButtonText?: string,
  autoHeight?: boolean,
  chunkClickable?: boolean,
  chunkMiniMap?: boolean,
  colorDetailPageRoot?: (Color) => string | string,
  colorWallBgColor: string,
  colorWallPageRoot?: (string | Color) => string | string,
  displayAddButton?: boolean,
  displayAddButtonText?: boolean,
  displayDetailsLink?: boolean,
  displayInfoButton?: boolean,
  hiddenSections?: string[],
  inactiveColorRouteBuilderRef?: { current: (Color) => void },
  isChipLocator?: boolean,
  loading: boolean,
  leftHandDisplay: boolean,
  routeType: string,
  swatchMaxSize: number,
  swatchMinSize: number,
  swatchMinSizeZoomed: number,
  swatchSizeZoomed: number,
  updateA11y: Function
}

export const colorWallContextDefault: ColorWallContextProps = {
  activeColorRouteBuilderRef: undefined,
  addButtonText: undefined,
  autoHeight: false,
  chunkClickable: false,
  chunkMiniMap: false,
  colorDetailPageRoot: undefined,
  colorWallBgColor: '#EEEEEE',
  colorWallPageRoot: undefined,
  displayAddButton: false,
  displayAddButtonText: false,
  displayDetailsLink: true,
  displayInfoButton: false,
  hiddenSections: [],
  inactiveColorRouteBuilderRef: undefined,
  isChipLocator: false,
  loading: false,
  leftHandDisplay: false,
  routeType: 'hash',
  swatchMaxSize: 33,
  swatchMinSize: 14,
  swatchMinSizeZoomed: 50,
  swatchSizeZoomed: 50,
  updateA11y: noop,
  viewColorText: false
}

const ColorWallContext = React.createContext<Object>({
  ...colorWallA11yContextDefault,
  ...colorWallContextDefault
})

export default ColorWallContext
