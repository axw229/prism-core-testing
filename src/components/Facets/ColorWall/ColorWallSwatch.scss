.color-swatch-inner {
  $flower-size-scale: 2;
  $flower-distance-scale: .35;

  $bloom-activate-delay: 0s;
  $bloom-activate-duration: .2s;
  $bloom-activate-z-duration: $bloom-activate-delay + $bloom-activate-duration;

  $bloom-deactivate-delay: .0s;
  $bloom-deactivate-duration: .2s;
  $bloom-deactivate-z-duration: $bloom-deactivate-delay + $bloom-deactivate-duration;

  @mixin getRaise($scale, $amt, $z, $i: '') {
    $this-scale: $amt * $scale;

    width: $this-scale unquote($i);
    height: $this-scale unquote($i);
    margin: round($this-scale / -2 + 50%) 0 0 round($this-scale / -2 + 50%) unquote($i);
    z-index: $z unquote($i);
    box-shadow: 0 0 0 1px white, 0 0 4px rgba(black, .25);
  }

  @keyframes fadeIn {
    0% {
      opacity: 0;
    }
    100% {
      opacity: 1;
    }
  }

  border: 1px solid #fff;
  height: 100%;
  width: 100%;
  top: 0;
  left: 0;
  position: absolute;
  will-change: top, left, transform, width, height, margin, box-shadow, font-size, z-index;
  /* autoprefixer: ignore next */
  transition-property: top, left, transform, width, height, margin, z-index, box-shadow;

  transition-delay: $bloom-deactivate-delay;

  transition-duration:
    $bloom-deactivate-duration,
    $bloom-deactivate-duration,
    $bloom-deactivate-duration,
    $bloom-deactivate-duration,
    $bloom-deactivate-duration,
    $bloom-deactivate-duration,
    $bloom-deactivate-z-duration,
    $bloom-deactivate-duration;

  transition-timing-function: ease-out;
  z-index: 1;
  color: rgba(black, .9);
  cursor: pointer;
  box-shadow: 0 0 0 1px white, 0 0 0px rgba(black, .25);

  &:hover {
    @include getRaise(1, 140%, 100);
    transition-duration: 0s;
    transition-delay: .1s;
  }

  &:focus {
    @include getRaise(1, 160%, 200);
    transition-duration: 0s;
    transition-delay: 0s;
  }

  &__content {
    position: relative;
    font-size: .7rem;
    padding: .5em;
    animation: fadeIn $bloom-activate-duration ease-in-out ($bloom-activate-duration + $bloom-activate-delay) * 1.25 both;
    text-align: left;
    height: 100%;
    line-height: 1;
    overflow: hidden;

    @media (min-width: 1400px) {
      font-size: .9rem;
    }

    &__number,
    &__name {
      margin: 0;
    }

    &__name {
      font-weight: bold;
    }

    &__number {
      opacity: .6;
      font-size: .8em;
      margin-bottom: .4em;
    }

    &__add,
    &__details {
      position: absolute;
      bottom: .5em;
      color: inherit;
      cursor: pointer;
      background: none;
      border: none;
      margin: 0;
      box-shadow: none;
      opacity: .6;
      transition: all .2s;
      text-align: center;

      &:hover, &:focus {
        opacity: 1;
      }
    }

    &__add {
      left: .5em;
    }

    &__details {
      right: .5em;
    }
  }

  &--dark-color {
    color: rgba(white, .9);
  }

  &--bloom {
    /* autoprefixer: ignore next */
    transition-property: top, left, transform, width, height, margin, z-index, box-shadow;

    transition-delay: 0s !important;

    transition-duration:
      $bloom-activate-duration,
      $bloom-activate-duration,
      $bloom-activate-duration,
      $bloom-activate-duration,
      $bloom-activate-duration,
      $bloom-activate-duration,
      $bloom-activate-z-duration,
      $bloom-activate-duration !important;
  }

  &--bloom-lvl {
    &-0 {
      @include getRaise($flower-size-scale, 200%, 800, !important);
      transition-duration: 0s !important;
      cursor: default;
    }

    &-n0-5 {
      @include getRaise($flower-size-scale, 180%, 750, !important);

      &:hover, &:focus {
        @include getRaise($flower-size-scale, 195%, 750, !important);
      }
    }

    &-n1 {
      @include getRaise($flower-size-scale, 165%, 700, !important);

     &:hover, &:focus {
        @include getRaise($flower-size-scale, 180%, 700, !important);
      }
    }

    &-n1-5 {
      @include getRaise($flower-size-scale, 160%, 650, !important);

      &:hover, &:focus {
        @include getRaise($flower-size-scale, 175%, 650, !important);
      }
    }

    &-n2 {
      @include getRaise($flower-size-scale, 130%, 600, !important);

      &:hover, &:focus {
        @include getRaise($flower-size-scale, 155%, 600, !important);
      }
    }
  }

  &--offset {
    &-x {
      &-n2 {
        top: -150% * $flower-distance-scale;
        &--primary {
          top: -166% * $flower-distance-scale;
        }
      }
      &-n1 {
        top: -100% * $flower-distance-scale;
        &--primary {
          top: -125% * $flower-distance-scale;
        }
      }
      &-1 {
        top: 100% * $flower-distance-scale;
        &--primary {
          top: 125% * $flower-distance-scale;
        }
      }
      &-2 {
        top: 150% * $flower-distance-scale;
        &--primary {
          top: 166% * $flower-distance-scale;
        }
      }
    }

    &-y {
      &-n2 {
        left: -150% * $flower-distance-scale;
        &--primary {
          left: -166% * $flower-distance-scale;
        }
      }
      &-n1 {
        left: -100% * $flower-distance-scale;
        &--primary {
          left: -125% * $flower-distance-scale;
        }
      }
      &-1 {
        left: 100% * $flower-distance-scale;
        &--primary {
          left: 125% * $flower-distance-scale;
        }
      }
      &-2 {
        left: 150% * $flower-distance-scale;
        &--primary {
          left: 166% * $flower-distance-scale;
        }
      }
    }
  }
}