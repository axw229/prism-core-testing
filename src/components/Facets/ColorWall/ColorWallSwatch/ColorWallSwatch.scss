.color-swatch-inner {
  $_self: &;

  $flower-size-scale: 1.5;
  $flower-distance-scale: .15;

  $bloom-activate-delay: getVar('colorWall.swatchActivateDelayMS') * 1ms;
  $bloom-activate-duration: getVar('colorWall.swatchActivateDurationMS') * 1ms;
  $bloom-activate-z-duration: $bloom-activate-delay + $bloom-activate-duration;

  $bloom-deactivate-delay: getVar('colorWall.swatchDeactivateDelayMS') * 1ms;
  $bloom-deactivate-duration: getVar('colorWall.swatchDeactivateDurationMS') * 1ms;
  $bloom-deactivate-z-duration: $bloom-deactivate-delay + $bloom-deactivate-duration;

  $swatch-scale-amt: 1.25;
  $swatch-scale-amt--zoomed-in: 1.1;

  @mixin getRaise($scale, $amt, $z, $i: '') {
    $this-scale: $amt * $scale;

    width: $this-scale unquote($i);
    height: $this-scale unquote($i);
    margin: round($this-scale / -2 + 50%) 0 0 round($this-scale / -2 + 50%) unquote($i);
    z-index: $z unquote($i);
    box-shadow: 0 0 0 1px white, 0 0 10px rgba(black, .25) !important;
  }

  @mixin getEdgeCompensation($scale, $amt, $i: '') {
    $this-scale: $amt * $scale;

    &#{$_self}--compensate-x-n2 {
      left: $this-scale / -3 unquote($i);
    }

    &#{$_self}--compensate-x-n1 {
      left: $this-scale / -6 unquote($i);
    }

    &#{$_self}--compensate-x-1 {
      left: $this-scale / 6 unquote($i);
    }

    &#{$_self}--compensate-x-2 {
      left: $this-scale / 3 unquote($i);
    }

    &#{$_self}--compensate-y-n2 {
      top: $this-scale / -3 unquote($i);
    }

    &#{$_self}--compensate-y-n1 {
      top: $this-scale / -6 unquote($i);
    }

    &#{$_self}--compensate-y-1 {
      top: $this-scale / 6 unquote($i);
    }

    &#{$_self}--compensate-y-2 {
      top: $this-scale / 3 unquote($i);
    }
  }

  @mixin getFocusScaling($scale) {
    &:hover, &:focus {
      transform: scale($scale);
    }

    &#{$_self}--pos-t {
      &:hover, &:focus {
        transform: scale($scale) translateY(percentage(($scale - 1) / $scale / 2));
      }
    }

    &#{$_self}--pos-tl {
      &:hover, &:focus {
        transform: scale($scale) translateY(percentage(($scale - 1) / $scale / 2)) translateX(percentage(($scale - 1) / $scale / 2));
      }
    }

    &#{$_self}--pos-tr {
      &:hover, &:focus {
        transform: scale($scale) translateY(percentage(($scale - 1) / $scale / 2)) translateX(percentage((1 - $scale) / $scale / 2));
      }
    }

    &#{$_self}--pos-b {
      &:hover, &:focus {
        transform: scale($scale) translateY(percentage((1 - $scale) / $scale / 2));
      }
    }

    &#{$_self}--pos-bl {
      &:hover, &:focus {
        transform: scale($scale) translateY(percentage((1 - $scale) / $scale / 2)) translateX(percentage(($scale - 1) / $scale / 2));
      }
    }

    &#{$_self}--pos-br {
      &:hover, &:focus {
        transform: scale($scale) translateY(percentage((1 - $scale) / $scale / 2)) translateX(percentage((1 - $scale) / $scale / 2));
      }
    }

    &#{$_self}--pos-l {
      &:hover, &:focus {
        transform: scale($scale) translateX(percentage(($scale - 1) / $scale / 2));
      }
    }

    &#{$_self}--pos-r {
      &:hover, &:focus {
        transform: scale($scale) translateX(percentage((1 - $scale) / $scale / 2));
      }
    }
  }

  @keyframes fadeIn {
    0% {
      opacity: 0;
    }
    100% {
      opacity: 1;
    }
  }

  border: 1px solid #fff;
  height: 100%;
  width: 100%;
  top: 0;
  left: 0;
  position: absolute;
  font-size: $base;
  z-index: 1;

  &--clickable {
    cursor: pointer;

    &:hover, &:focus {
      z-index: 100;
      transition: none;
    }

    .color-wall-swatch-list--show-all & {
      @include getFocusScaling($swatch-scale-amt);
    }

    .color-wall-swatch-list--zoomed & {
      @include getFocusScaling($swatch-scale-amt--zoomed-in);
    }
  }

  &--dynamic {
    will-change: top, left, transform, width, height, margin, box-shadow, font-size, z-index;
    /* autoprefixer: ignore next */
    transition-property: top, left, transform, width, height, margin, z-index, box-shadow;

    transition-delay: $bloom-deactivate-delay;

    transition-duration:
      $bloom-deactivate-duration,
      $bloom-deactivate-duration,
      $bloom-deactivate-duration,
      $bloom-deactivate-duration,
      $bloom-deactivate-duration,
      $bloom-deactivate-duration,
      $bloom-deactivate-z-duration,
      $bloom-deactivate-duration;

    transition-timing-function: ease-out;
    color: $color-near-black;
    box-shadow: 0 0 0 1px white, 0 0 0px rgba(black, .25) !important;
  }

  &__content {
    position: relative;
    padding: .5em;
    animation: fadeIn $bloom-activate-duration ease-in-out ($bloom-activate-duration + $bloom-activate-delay) * 1.25 both;
    text-align: left;
    height: 100%;
    line-height: 1;
    overflow: hidden;

    &__number,
    &__name {
      margin: 0;
    }

    &__name {
      font-weight: bold;
    }

    &__number {
      opacity: $color-near-opacity;
      font-size: .8em;
      margin-bottom: .4em;
    }

    &__add,
    &__details,
    &__info {
      position: absolute;
      bottom: 1em;
      color: inherit;
      cursor: pointer;
      background: none;
      border: none;
      margin: 0;
      box-shadow: none;
      opacity: $color-near-opacity;
      transition: all .2s;
      text-align: center;

      &:hover, &:focus {
        opacity: 1;
      }
    }

    &__add,
    &__details {
      left: .5em;
    }

    &__info {
      right: .5em;
    }
  }

  &--dark-color {
    color: rgba(white, .9);
  }

  &--bloom {
    /* autoprefixer: ignore next */
    transition-property: top, left, transform, width, height, margin, z-index, box-shadow;

    transition-delay: 0s !important;

    transition-duration:
      $bloom-activate-duration,
      $bloom-activate-duration,
      $bloom-activate-duration,
      $bloom-activate-duration,
      $bloom-activate-duration,
      $bloom-activate-duration,
      $bloom-activate-z-duration,
      $bloom-activate-duration !important;

    &:hover, &:focus {
      transform: unset !important;
    }
  }

  &--bloom-lvl {
    &-0 {
      @include getRaise($flower-size-scale, 200%, 900, !important);
      @include getEdgeCompensation($flower-size-scale, 200%, !important);
      cursor: default;
    }

    &-n0-5 {
      @include getRaise($flower-size-scale, 180%, 750, !important);
      @include getEdgeCompensation($flower-size-scale, 180%, !important);
    }

    &-n1 {
      @include getRaise($flower-size-scale, 165%, 700, !important);
      @include getEdgeCompensation($flower-size-scale, 165%, !important);
    }

    &-n1-5 {
      @include getRaise($flower-size-scale, 160%, 650, !important);
      @include getEdgeCompensation($flower-size-scale, 160%, !important);
    }

    &-n2 {
      @include getRaise($flower-size-scale, 130%, 600, !important);
      @include getEdgeCompensation($flower-size-scale, 130%, !important);
    }
  }

  &__engage-link {
    height: 100%;
    left: 0;
    position: absolute;
    top: 0;
    width: 100%;
  }
}