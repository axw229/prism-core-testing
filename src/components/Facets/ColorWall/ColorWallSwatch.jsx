/* eslint-disable */
// @flow
import React, { PureComponent, Fragment } from 'react'

import { type Color } from '../../../shared/types/Colors'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

type Props = {
  color: Color,
  onEngage: Function,
  onAdd: Function,
  level?: number,
  offsetX?: number,
  offsetY?: number,
  active?: boolean
}

class ColorWallSwatch extends PureComponent<Props> {
  ref: any = null

  static getComputedValue (el: any, styleProp: string) {
    if (window.getComputedStyle) {
      return window.getComputedStyle(el, null).getPropertyValue(styleProp);
    } else if (el.currentStyle) {
      return el.currentStyle[styleProp];
    }
  }

  constructor (props: Props) {
    super(props)

    this.handleAddClick = this.handleAddClick.bind(this)
    this.handleDetailClick = this.handleDetailClick.bind(this)
    this.handleSwatchClick = this.handleSwatchClick.bind(this)
  }

  getOuterClasses () {
    const { level, offsetX, offsetY, color, active } = this.props
    let classes = ['inner', 'color-swatch-inner']

    if (active) {
      classes.push('color-swatch-inner--active')
    }

    if (level) {
      classes.push('color-swatch-inner--bloom')
      classes.push('color-swatch-inner--lvl-' + level.toString().replace(/\./g, '-'))
    }

    if (offsetX) {
      let className = 'color-swatch-inner--offset-x-' + offsetX.toString().replace(/-/g, 'n')
      classes.push(className)

      if (!offsetY) {
        classes.push(className + '--primary')
      }
    }

    if (offsetY) {
      let className = 'color-swatch-inner--offset-y-' + offsetY.toString().replace(/-/g, 'n')
      classes.push(className)

      if (!offsetX) {
        classes.push(className + '--primary')
      }
    }

    if (!offsetX && !offsetY && level) {
      classes.push('color-swatch-inner--bloom-center')
    }

    if (color.isDark) {
      classes.push('color-swatch-inner--dark-color')
    }

    return classes.join(' ')
  }

  getOuterStyles () {
    let styleObj: {
      zIndex?: number
    } = {}

    if (this.ref) {
      styleObj.zIndex = ColorWallSwatch.getComputedValue(this.ref, 'z-index')
    }

    return styleObj
  }

  getInnerStyles () {
    const { color } = this.props

    let styleObj: {
      backgroundColor: string
    } = {
      backgroundColor: color.hex
    }

    return styleObj
  }

  render () {
    const { active, level, color } = this.props

    return (
      <React.Fragment>
        <li className='color-wall-swatches__swatch'>
          <div className={this.getOuterClasses()}
            ref={(el) => this.ref = el}
            style={this.getOuterStyles()}>
            {level === 3 && (
              <div className='color-swatch-inner__color-container' style={this.getInnerStyles()}>
                <div className='color-swatch-inner__content'>
                  <p>{`${color.brandKey ? color.brandKey + ' ' : ''} ${color.colorNumber}`}</p>
                  <p><strong>{color.name}</strong></p>
                  <button autoFocus onClick={this.handleAddClick} className='color-swatch-inner__content__add'><FontAwesomeIcon icon='plus' size='1x' /></button>
                  <button onClick={this.handleDetailClick} className='color-swatch-inner__content__details'><FontAwesomeIcon icon='info' size='1x' /></button>
                </div>
              </div>
            ) || (
              <button onClick={this.handleSwatchClick}
                className='color-swatch-inner__color-container'
                style={this.getInnerStyles()}
                />
            )}
          </div>
        </li>
      </React.Fragment>
    )
  }

  handleSwatchClick = function handleSwatchClick () {
    this.props.onEngage(this.props.color)
  }

  handleAddClick = function handleAddClick() {
    this.props.onAdd(this.props.color)
  }

  handleDetailClick = function handleDetailClick() {
    const { color } = this.props
    alert(`Color details for ${color.name}`)
  }
}

export default ColorWallSwatch
