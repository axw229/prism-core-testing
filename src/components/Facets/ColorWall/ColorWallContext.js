// @flow
import React from 'react'
import { type GridBounds } from './ColorWall.flow'
import noop from 'lodash/noop'

export type ColorWallA11yContextProps = {
  a11yFocusCell: number[] | void,
  a11yFocusChunk: GridBounds,
  a11yFromKeyboard: boolean,
  a11yFocusOutline: boolean,
  a11yMaintainFocus: boolean
}

export const colorWallA11yContextDefault: ColorWallA11yContextProps = {
  a11yFocusCell: undefined,
  a11yFocusChunk: undefined,
  a11yFromKeyboard: false,
  a11yFocusOutline: false,
  a11yMaintainFocus: false
}

export type ColorWallContextProps = {
  addButtonText?: string,
  chunkClickable?: boolean,
  chunkMiniMap?: boolean,
  colorDetailPageRoot?: string,
  colorNumOnBottom?: boolean,
  colorWallBgColor: string,
  colorWallChunkPageRoot?: string,
  displayAddButton?: boolean,
  displayAddButtonText?: boolean,
  displayDetailsLink?: boolean,
  displayInfoButton?: boolean,
  hiddenSections?: string[],
  loading: boolean,
  swatchMaxSize: number,
  swatchSizeZoomed: number,
  swatchMinSize: number,
  swatchMinSizeZoomed: number,
  updateA11y: Function
}

export const colorWallContextDefault: ColorWallContextProps = {
  addButtonText: undefined,
  chunkClickable: false,
  chunkMiniMap: false,
  colorDetailPageRoot: undefined,
  colorNumOnBottom: false,
  colorWallChunkPageRoot: undefined,
  colorWallBgColor: '#EEEEEE',
  displayAddButton: false,
  displayAddButtonText: false,
  displayDetailsLink: true,
  viewColorText: false,
  displayInfoButton: false,
  hiddenSections: [],
  loading: false,
  swatchMaxSize: 33,
  swatchSizeZoomed: 50,
  swatchMinSize: 14,
  swatchMinSizeZoomed: 50,
  updateA11y: noop
}

const ColorWallContext = React.createContext<Object>({
  ...colorWallA11yContextDefault,
  ...colorWallContextDefault
})

export default ColorWallContext
