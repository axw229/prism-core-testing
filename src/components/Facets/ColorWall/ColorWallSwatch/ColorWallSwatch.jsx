// @flow
import React, { useMemo, useEffect } from 'react'
import { Link } from 'react-router-dom'
import kebabCase from 'lodash/kebabCase'
import { type Color } from '../../../../shared/types/Colors'
import { numToAlphaString, arrayToSpacedString } from '../../../../shared/helpers/StringUtils'
import { fullColorName, fullColorNumber } from '../../../../shared/helpers/ColorUtils'
import { CLASS_NAMES } from './shared'
import ColorWallContext, { type ColorWallContextProps } from 'src/components/Facets/ColorWall/ColorWallContext'
import './ColorWallSwatch.scss'
import 'src/scss/convenience/visually-hidden.scss'
import OmniButton from './ColorWallSwatchButtons/OmniButton'
import { generateColorDetailsPageUrl } from 'src/shared/helpers/ColorUtils'
import { FormattedMessage } from 'react-intl'

type Props = {
  color: Color,
  focus?: boolean,
  thisLink?: string,
  showContents?: boolean,
  onAdd?: Function,
  onClick?: Function,
  level?: number,
  compensateX?: number,
  compensateY?: number,
  active?: boolean
}

const ColorWallSwatch = React.forwardRef<Props, any>((props: Props, ref: any) => {
  const { onAdd, onClick, showContents, color, thisLink, focus, level, active, compensateX, compensateY } = props
  const { displayAddButton, displayInfoButton, displayDetailsLink, colorDetailPageRoot }: ColorWallContextProps = React.useContext(ColorWallContext)
  const handleOnAdd = useMemo(() => {
    return () => {
      if (onAdd) {
        onAdd(color)
      }
    }
  }, [color, onAdd])

  const handleOnClick = useMemo(() => {
    return (e) => {
      if (onClick) {
        onClick(e)
      }
    }
  }, [onClick])

  const classes = useMemo(() => {
    let classes = [ CLASS_NAMES.BASE, CLASS_NAMES.BASE_DYNAMIC ]
    if (color.isDark) { classes.push(CLASS_NAMES.BASE_DARK) }
    if (active) { classes.push(CLASS_NAMES.BASE_ACTIVE) }
    // if we have a link to this swatch's active URL assume it's a clickable swatch
    if (thisLink) { classes.push(CLASS_NAMES.BASE_CLICKABLE) }
    if (!isNaN(level)) {
      classes.push(CLASS_NAMES.BASE_BLOOM)
      classes.push(CLASS_NAMES.BASE_BLOOM_LVL_ + numToAlphaString(level))
      if (compensateX) { classes.push(CLASS_NAMES.BASE_COMPENSATE_X_ + numToAlphaString(compensateX)) }
      if (compensateY) { classes.push(CLASS_NAMES.BASE_COMPENSATE_Y_ + numToAlphaString(compensateY)) }
    }
    if (focus) { classes.push(CLASS_NAMES.BASE_FOCUS) }

    return arrayToSpacedString(classes)
  }, [color, active, thisLink, level, compensateX, compensateY, focus])

  const { content, refData } = useMemo<any, Object | void>(() => {
    if (!showContents) {
      const inner = (
        <span className='visually-hidden'>{fullColorName(color.brandKey, color.colorNumber, color.name)}</span>
      )

      if (thisLink) {
        return {
          content: (
            <Link to={thisLink} className={CLASS_NAMES.ENGAGE_LINK} onClick={handleOnClick} tabIndex={-1}>
              {inner}
            </Link>
          ),
          refData: {
            internalLink: thisLink,
            onClick: onClick
          }
        }
      } else {
        return {
          content: inner
        }
      }
    }

    if (displayAddButton) {
      return {
        content: (
          <OmniButton
            icon={OmniButton.ICONS.ADD}
            onClick={handleOnAdd}
            className={`${CLASS_NAMES.CONTENT_CTA} ${CLASS_NAMES.CONTENT_CTA_L}`}
            tabIndex={-1}
          />
        ),
        refData: {
          onClick: onClick
        }
      }
    } else if (displayInfoButton) {
      const link = generateColorDetailsPageUrl(color)
      return {
        content: (
          <OmniButton
            link={link}
            icon={OmniButton.ICONS.INFO}
            className={`${CLASS_NAMES.CONTENT_CTA} ${CLASS_NAMES.CONTENT_CTA_R}`}
            tabIndex={-1}
          />
        ),
        refData: {
          internalLink: link
        }
      }
    } else if (displayDetailsLink) {
      const link = colorDetailPageRoot ? `${colorDetailPageRoot}/${color.brandKey}${color.colorNumber}-${kebabCase(color.name)}` : generateColorDetailsPageUrl(color)
      return {
        content: (
          <OmniButton
            link={link}
            className={`${CLASS_NAMES.CONTENT_CTA} ${CLASS_NAMES.CONTENT_CTA_L} ${focus ? CLASS_NAMES.CONTENT_CTA_FOCUS : ''}`}
            tabIndex={-1}>
            <FormattedMessage id='VIEW_DETAILS' />
          </OmniButton>
        ),
        refData: {
          [colorDetailPageRoot ? 'externalLink' : 'internalLink']: link
        }
      }
    }

    return {}
  }, [color, onClick, thisLink, handleOnAdd, displayAddButton, displayInfoButton, displayDetailsLink, colorDetailPageRoot, showContents])

  useEffect(() => {
    if (ref) {
      ref.current = refData
    }
  }, [refData, ref])

  return (
    <div className={CLASS_NAMES.SWATCH}>
      <div className={classes} style={{ background: color.hex }}>
        {showContents ? (
          <section className={CLASS_NAMES.CONTENT}>
            <p className={CLASS_NAMES.CONTENT_NUMBER}>{fullColorNumber(color.brandKey, color.colorNumber)}</p>
            <p className={CLASS_NAMES.CONTENT_NAME}>{color.name}</p>
            {content}
          </section>
        ) : (
          content
        )}
      </div>
    </div>
  )
})

export default ColorWallSwatch
