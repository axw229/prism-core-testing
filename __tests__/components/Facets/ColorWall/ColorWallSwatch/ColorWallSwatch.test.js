/* eslint-env jest */
import React from 'react'
import { shallow, mount } from 'enzyme'
import ColorWallSwatch from 'src/components/Facets/ColorWall/ColorWallSwatch/ColorWallSwatch'
import * as Colors from '__mocks__/data/color/Colors'
import { Link } from 'react-router-dom'
import ColorWallContext from 'src/components/Facets/ColorWall/ColorWallContext'
import { fullColorName } from 'src/shared/helpers/ColorUtils'
import AddButton from 'src/components/Facets/ColorWall/ColorWallSwatch/ColorWallSwatchButtons/AddButton'

const color = Colors.getColor()
const clickFn = jest.fn()

const defaultProps = {
  color: color,
  focus: false,
  thisLink: '',
  detailsLink: '',
  showContents: false,
  onAdd: clickFn,
  onClick: clickFn,
  active: false
}

const getColorWallSwatch = (props = {}) => {
  return shallow(<ColorWallSwatch {...defaultProps} {...props} />)
}
const getDeepMountColorWallSwatch = (props = {}) => {
  const config = { displayInfoButton: true }
  return mount(<ColorWallContext.Provider value={config}>
    <ColorWallSwatch {...defaultProps} {...props} />
  </ColorWallContext.Provider>)
}

describe('snapshot test', () => {
  const ColorWallSwatch = getColorWallSwatch()
  it('ColorWallSwatch should match the Snapshot', () => {
    expect(ColorWallSwatch).toMatchSnapshot()
  })
})

describe('rendering test', () => {
  it('ColorWallSwatch should render color title as default', () => {
    expect(typeof fullColorName(color.brandKey, color.colorNumber, color.name)).toBe('string')
  })
  it('ColorWallSwatch should rendering buttons correctly when props showContents equal to true', () => {
    const ColorWallSwatch = getColorWallSwatch({ showContents: true })
    expect(ColorWallSwatch.find(ColorWallContext.Consumer).exists()).toBe(true)
  })
  it('ColorWallSwatch should rendering Link correctly when props thislink is not equal to null', () => {
    const ColorWallSwatch = getColorWallSwatch({ thisLink: 'test' })
    expect(ColorWallSwatch.find(Link).exists()).toBe(true)
  })
})

describe('event test', () => {
  it('AddButton should working fine', () => {
    const ColorWallSwatch = getDeepMountColorWallSwatch({ showContents: true })
    ColorWallSwatch.find(AddButton).props().onClick()
    expect(clickFn).toHaveBeenCalled()
  })
  it('Link component should work fine', () => {
    const ColorWallSwatch = getColorWallSwatch({ thisLink: 'test' })
    ColorWallSwatch.find(Link).simulate('click')
    expect(clickFn).toHaveBeenCalled()
  })
})
