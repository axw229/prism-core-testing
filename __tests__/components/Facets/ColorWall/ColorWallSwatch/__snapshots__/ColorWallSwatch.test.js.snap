// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test ColorWallSwatch should match the Snapshot 1`] = `
<div
  className="color-wall-swatches__swatch"
>
  <div
    className="color-swatch-inner color-swatch-inner--dynamic "
    style={
      Object {
        "background": "#ebd1db",
      }
    }
  >
    <span
      className="visually-hidden"
    >
      SW 6561 Teaberry
    </span>
  </div>
</div>
`;
